New PerformanceObserver tests

JosephPecoraro opened the pull request:
- Test sort order of entries in PerformanceObserverEntryList
- Test behavior of disconnect on observer with entries in buffer
- Test behavior of modifying an Observer in its callback
- Test types of parameters and this value in callback

<!-- Reviewable:start -->

<!-- Reviewable:end -->

JosephPecoraro commented on performance-timeline/po-observe.html:
```diff
@@ -40,9 +40,29 @@
+    var finish = t.step_func(function () { t.done(); });
+    var observer = new PerformanceObserver(  
+      function (entryList, obs) {
+        var self = this;
```
I filed https://github.com/w3c/testharness.js/issues/236 to make this better. Once that is resolved we can use a less ugly approach here.

JosephPecoraro commented:
This adds tests for https://github.com/w3c/performance-timeline/issues/65, https://github.com/w3c/performance-timeline/issues/66,  https://github.com/w3c/performance-timeline/issues/67.

This uncovers a few issues in Firefox's implementation (sort + callback observe). As well as makes tests match the spec for disconnect behavior, which does not match Chrome's behavior. The this value checking also uncovers an issue in WebKit's [coming implementation](https://bugs.webkit.org/show_bug.cgi?id=167546).

JosephPecoraro commented:
Is there anything else I should do for this?

tobie commented:
Ping @wpt-pr-bot.

igrigorik commented:
/cc @spanicker for review :)

tobie commented:
w3c-test:mirror

tobie commented:
And, I just realized I'm not a collaborator on this repo, so that won't help. :(

BTW, you can add folks to the [OWNERS file](https://github.com/w3c/web-platform-tests/blob/master/performance-timeline/OWNERS) if you want them to be pinged by the review bot.

spanicker commented on performance-timeline/po-entries-sort.html:
```diff
@@ -0,0 +1,77 @@
+          stored_entries_by_name = entryList.getEntriesByName("name-repeat");
+
+          checkSorted(stored_entries);
+          checkEntriesInOrder(stored_entries, [
```
wouldn't checkEntries suffice here? whether the set of entries is sorted by startTime was already checked in checkSorted. In terms of ordering - ordering by startTime is the only guarantee that Performance Observer makes AFAICT.

JosephPecoraro replied:
Please see: https://github.com/w3c/performance-timeline/issues/67

You are correct that startTime is the only guarantee. If we really agree that is the only guarantee that should be made, then I can reduce this.

spanicker replied:
From Chrome's perspective requiring any other guarantees would be problematic (especially going forward).

JosephPecoraro replied:
Based on the 2/8 meeting this test should be less strict. Namely just guarantee start time sorted, no guaranteed order for entries with equivalent start times.

spanicker commented on performance-timeline/po-disconnect.html:
```diff
@@ -30,15 +30,16 @@
-            [{ entryType: "mark", name: "mark1"}]);
-          t.done();
+        t.step_func(function (entryList, obs) {
+          assert_unreached("This callback must not be invoked");
```
isn't it plausible that the callback is invoked between calling performance.mark("mark1") and observer.disconnect() (Although the original test did seem problematic)

JosephPecoraro replied:
I don't believe it is plausible. The spec says to queue a task when the mark is created, which I believe is an asynchronous operation. By the time the task runs, the observer has cleared its individual buffer via disconnect, so it doesn't have anything to deliver. If tasks are not async operations, then may could be right.

See my complete diagnosis on: https://github.com/w3c/performance-timeline/issues/66

spanicker replied:
I commented on that issue bug, we had discussed this before in https://crbug.com/607324 and the test was deemed appropriate at the time.

spanicker replied:
Wouldn't this change break for chrome today?
I know we discussed updating the chrome implementation if there aren't too many callers depending on this, and updating the spec with takeRecords, so this change seems pre-mature (and would prevent importing these tests into chromium's auto-sync process for WPT)
https://github.com/w3c/performance-timeline/issues/66#issuecomment-277830576
@igrigorik WDYT?

JosephPecoraro replied:
Correct, Chrome would fail this particular test. However, this test matches both the spec and Firefox's implementation today. I'm fine with waiting until the spec is updated; it it goes the direction of adding takeRecords, I suspect this test will still be valid.

Chromium's auto-sync process not importing failing tests sounds like an issue / policy to be solved by the chromium project. Inevitably there will be correct web-platform-tests that a browser will fail to pass. I'm not sure that concern is relevant to the review of adding new tests to web-platform-tests if the tests themselves are correct. That said, I am new to this process.

spanicker replied:
Agreed that the test process is a separate thing from current issue.
I'll let Ilya / Todd chime in here on whether we should update the test (and break test for chrome as a side effect) or wait until spec update for takeRecords etc.

igrigorik replied:
/cc @RByers for guidance on WPT bits.. I agree with @JosephPecoraro that this is a more general (good) problem, that we'll also have to solve.

plehegar replied:
if those tests valid, shouldn't we merge them?

JosephPecoraro replied:
I believe these tests are still valid. WebKit still runs them.

spanicker replied:
Merging SGTM.

toddreifsteck commented:
I'll defer to @spanicker and @JosephPecoraro to sort things out on this review as both are implementers of the specification. Once both agree that all spec updates are made and tests are solid, we can merge.

JosephPecoraro commented on performance-timeline/performanceobservers.js:
```diff
@@ -19,3 +19,41 @@ function checkEntries(perfEntriesToCheck, expectedEntries) {
+    if (!pe)
+      continue;
+    assert_equals(ex.name, pe.name, "entry name must match");
+    assert_equals(ex.entryType, pe.entryType, "entry entryType must match");
```
I'm not sure what you mean. This method (checkEntriesInOrder) asserts name/type of each entry. The other method (checkEntries) only asserts entries exist with given names/types. Those are similar but different operations.

JosephPecoraro commented:
Updated the sort order test. It now only guarantees startTime sort, and Chrome now passes it. This can be re-reviewed.

w3c-bots commented:
# Chrome (unstable channel) #
Testing web-platform-tests at revision 5dadf4b127e2cb0e220863e21933cfada2c348b3
Using browser at version 58.0.3004.3 dev
Starting 10 test iterations
All results were stable
## All results ##
<details>
<summary><a href="http://w3c-test.org/submissions/4660/performance-timeline/po-callback-mutate.html">/performance-timeline/po-callback-mutate.html</a></summary>


|                                             Subtest                                              | Results |
|--------------------------------------------------------------------------------------------------|---------|
|                                                                                                  | OK      |
| `PerformanceObserver modifications inside callback should update filtering and not clear buffer` | FAIL    |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/4660/performance-timeline/po-disconnect.html">/performance-timeline/po-disconnect.html</a></summary>


|                            Subtest                            | Results |
|---------------------------------------------------------------|---------|
|                                                               | OK      |
| `An observer disconnected must not have its callback invoked` | FAIL    |
| `disconnected callbacks must not be invoked`                  | PASS    |
| `disconnecting an unconnected observer is a no-op`            | PASS    |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/4660/performance-timeline/case-sensitivity.html">/performance-timeline/case-sensitivity.html</a></summary>


|                   Subtest                    | Results |
|----------------------------------------------|---------|
|                                              | OK      |
| `getEntriesByName values are case sensitive` | PASS    |
| `getEntriesByType values are case sensitive` | PASS    |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/4660/performance-timeline/po-resource.html">/performance-timeline/po-resource.html</a></summary>


|              Subtest              | Results |
|-----------------------------------|---------|
|                                   | OK      |
| `resource entries are observable` | PASS    |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/4660/performance-timeline/po-navigation.html">/performance-timeline/po-navigation.html</a></summary>


|             Subtest              | Results |
|----------------------------------|---------|
|                                  | OK      |
| `navigation entry is observable` | PASS    |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/4660/performance-timeline/po-observe.html">/performance-timeline/po-observe.html</a></summary>


|                                    Subtest                                    | Results |
|-------------------------------------------------------------------------------|---------|
|                                                                               | OK      |
| `replace observer if already present`                                         | PASS    |
| `Check observer callback parameter and this values`                           | PASS    |
| `Filter unsupported entryType entryType names within the entryTypes sequence` | PASS    |
| `no entryTypes throws a TypeError`                                            | PASS    |
| `Empty sequence entryTypes throws a TypeError`                                | PASS    |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/4660/performance-timeline/po-entries-sort.html">/performance-timeline/po-entries-sort.html</a></summary>


|                           Subtest                           | Results |
|-------------------------------------------------------------|---------|
|                                                             | OK      |
| `getEntries, getEntriesByType, getEntriesByName sort order` | PASS    |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/4660/performance-timeline/po-getentries.html">/performance-timeline/po-getentries.html</a></summary>


|                         Subtest                          | Results |
|----------------------------------------------------------|---------|
|                                                          | OK      |
| `getEntries, getEntriesByType and getEntriesByName work` | PASS    |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/4660/performance-timeline/po-mark-measure.html">/performance-timeline/po-mark-measure.html</a></summary>


|             Subtest              | Results |
|----------------------------------|---------|
|                                  | OK      |
| `entries are observable`         | PASS    |
| `measure entries are observable` | PASS    |
| `mark entries are observable`    | PASS    |

</details>

w3c-bots commented:
# Firefox (nightly channel) #
Testing web-platform-tests at revision 5dadf4b127e2cb0e220863e21933cfada2c348b3
Using browser at version BuildID 20170123125947; SourceStamp 36486fdc3813ef7943ae5b07b4128866d1938a6c
Starting 10 test iterations
All results were stable
## All results ##
<details>
<summary><a href="http://w3c-test.org/submissions/4660/performance-timeline/po-callback-mutate.html">/performance-timeline/po-callback-mutate.html</a></summary>


|                                             Subtest                                              | Results |
|--------------------------------------------------------------------------------------------------|---------|
|                                                                                                  | TIMEOUT |
| `PerformanceObserver modifications inside callback should update filtering and not clear buffer` | TIMEOUT |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/4660/performance-timeline/po-disconnect.html">/performance-timeline/po-disconnect.html</a></summary>


|                            Subtest                            | Results |
|---------------------------------------------------------------|---------|
|                                                               | OK      |
| `An observer disconnected must not have its callback invoked` | PASS    |
| `disconnected callbacks must not be invoked`                  | PASS    |
| `disconnecting an unconnected observer is a no-op`            | PASS    |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/4660/performance-timeline/case-sensitivity.html">/performance-timeline/case-sensitivity.html</a></summary>


|                   Subtest                    | Results |
|----------------------------------------------|---------|
|                                              | OK      |
| `getEntriesByName values are case sensitive` | PASS    |
| `getEntriesByType values are case sensitive` | PASS    |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/4660/performance-timeline/po-resource.html">/performance-timeline/po-resource.html</a></summary>


|              Subtest              | Results |
|-----------------------------------|---------|
|                                   | OK      |
| `resource entries are observable` | PASS    |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/4660/performance-timeline/po-navigation.html">/performance-timeline/po-navigation.html</a></summary>


|             Subtest              | Results |
|----------------------------------|---------|
|                                  | OK      |
| `navigation entry is observable` | FAIL    |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/4660/performance-timeline/po-observe.html">/performance-timeline/po-observe.html</a></summary>


|                                    Subtest                                    | Results |
|-------------------------------------------------------------------------------|---------|
|                                                                               | OK      |
| `replace observer if already present`                                         | PASS    |
| `Check observer callback parameter and this values`                           | PASS    |
| `Filter unsupported entryType entryType names within the entryTypes sequence` | PASS    |
| `no entryTypes throws a TypeError`                                            | PASS    |
| `Empty sequence entryTypes throws a TypeError`                                | PASS    |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/4660/performance-timeline/po-entries-sort.html">/performance-timeline/po-entries-sort.html</a></summary>


|                           Subtest                           | Results |
|-------------------------------------------------------------|---------|
|                                                             | OK      |
| `getEntries, getEntriesByType, getEntriesByName sort order` | FAIL    |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/4660/performance-timeline/po-getentries.html">/performance-timeline/po-getentries.html</a></summary>


|                         Subtest                          | Results |
|----------------------------------------------------------|---------|
|                                                          | OK      |
| `getEntries, getEntriesByType and getEntriesByName work` | FAIL    |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/4660/performance-timeline/po-mark-measure.html">/performance-timeline/po-mark-measure.html</a></summary>


|             Subtest              | Results |
|----------------------------------|---------|
|                                  | OK      |
| `entries are observable`         | PASS    |
| `measure entries are observable` | PASS    |
| `mark entries are observable`    | PASS    |

</details>

spanicker commented on performance-timeline/po-entries-sort.html:
```diff
@@ -0,0 +1,101 @@
+            {entryType: "mark", name: "name-repeat"},
+            {entryType: "mark", name: "name-repeat"},
+          ]);
+          checkEntriesInOrder(stored_entries, [
```
Why are these checks "checkEntriesInOrder" necessary, if you are already checking "checkSorted" earlier?
Maybe I'm missing something as I already asked this before:,

JosephPecoraro replied:
checkEntriesInOrder checks specific entry attributes at specific indexes. checkSorted only checked that the entries are sorted properly by startTime.

spanicker replied:
Sure I get that. In the public-web-perf call we agreed that we only guarantee sorting by startTime (as currently spec'd). So having a helper and tests to check for specific order (beyond startTime)  does not seem congruent to that.
See last comment in: 
https://github.com/w3c/performance-timeline/issues/67#issuecomment-278422238

JosephPecoraro replied:
For entries with equivalent startTime, while their order is not guaranteed there are other attributes we can check for that are guaranteed in the test case (like ensuring that they are "measure" and not "mark"). This part of the test is checking for everything that must be guaranteed.

Chrome Canary passes this test. But if you think this is adding extra complexity I'm fine with dropping it.

spanicker replied:
Yes this seems like unnecessary complexity and the addition of the helper "checkEntriesInOrder" is misleading also.
checkSorted and checkEntries should suffice for everything that performance-observer should guarantee. Otherwise let's fix them instead of adding overlapping helper methods.

igrigorik commented:
@JosephPecoraro apologies about the delay. Discussed this on the WG call today, and this looks good to merge.. Modulo merge conflicts. Any chance you'd be willing to push an update to resolve those?

JosephPecoraro commented:
I can take a look at updating this branch.

JosephPecoraro committed:
[performance-timeline] New PerformanceObserver Tests

- Test sort order of entries in PerformanceObserverEntryList
- Test behavior of disconnect on observer with entries in buffer
- Test behavior of modifying an Observer in its callback
- Test types of parameters and this value in callback

JosephPecoraro commented on performance-timeline/po-observe.any.js:
```diff
@@ -34,8 +34,29 @@
 
   async_test(function (t) {
   var observer = new PerformanceObserver(
+    var finish = t.step_func(function () { t.done(); });
```
This is missing a step_func. I'll be updating shortly now that I can test them.

igrigorik reviewed (approved):
👍

igrigorik commented:
@JosephPecoraro awesome work, thank you! Merging.
